File: 백준/Gold/1074. Z/Z.c
#include <stdio.h>

long long z_order(int N, int r, int c) {
    long long ans = 0;
    
    while (N > 0) {
        int size = 1 << (N - 1);  // 2^(N-1)
        
        if (r < size && c < size) {
            //  왼쪽 위: +0
        } else if (r < size && c >= size) {
            //  오른쪽 위
            ans += (long long)size * size;
            c -= size;
        } else if (r >= size && c < size) {
            //  왼쪽 아래
            ans += 2LL * size * size;
            r -= size;
        } else {
            // 오른쪽 아래
            ans += 3LL * size * size;
            r -= size;
            c -= size;
        }
        N--;
    }
    return ans;
}

int main() {
    int N, r, c;
    scanf("%d %d %d", &N, &r, &c);
    printf("%lld\n", z_order(N, r, c));
    return 0;
}

File: 백준/Gold/1074. Z/Z.py
def z_order(N, r, c):
    ans = 0
    
    while N > 0:
        size = 1 << (N - 1)  # 2^(N-1)
        
        if r < size and c < size:
            #  왼쪽 위: +0 (아무것도 안함)
            pass
        elif r < size and c >= size:
            #  오른쪽 위: +size*size
            ans += size * size
            c -= size
        elif r >= size and c < size:
            # 왼쪽 아래: +2*size*size
            ans += 2 * size * size
            r -= size
        else:
            # 오른쪽 아래: +3*size*size
            ans += 3 * size * size
            r -= size
            c -= size
        
        N -= 1
    
    return ans

# 입력 받기
N, r, c = map(int, input().split())
print(z_order(N, r, c))
